osu.Game/Overlays/Toolbar/Toolbar.cs

```diff
@@ -41,6 +41,9 @@ public class Toolbar : VisibilityContainer, IKeyBindingHandler<GlobalAction>
         // Toolbar and its components need keyboard input even when hidden.
         public override bool PropagateNonPositionalInputSubTree => true;
 
+        // IsHovered is used
+        public override bool HandlePositionalInput => true;
```

Comment:
Why does this override `HandlePositionalInput` but not actually handle any mouse events via overriding a handler, and why does `OsuFocusedOverlayContainer` compensate for that by checking the toolbar's `IsHovered`? I did this the way that seemed obvious, which is overriding

```csharp
protected override Handle(UIEvent e) => e is MouseEvent;
```

and removing the `IsHovered` checks in `OsuFocusedOverlayContainer` and it seems to work just as well...?

Reply:
This regresses a behavior overlays have.
<img width="1478" alt="Screen Shot 2021-08-29 at 1 47 32 PM" src="https://user-images.githubusercontent.com/35318437/131264929-685bbea5-6d8c-4c0f-b399-7ddfe0d7fae1.png">

Since the toolbar should be part of all overlays, a mouse down in the red area (i.e. dimmed area) and then a mouse up in the green area should not close the overlay. Will make a test to see what I mean, should've done that first.

Before:
![Screen Recording 2021-08-29 at 1 55 25 PM](https://user-images.githubusercontent.com/35318437/131265074-5b008c7e-23fd-429c-9eb0-5b3a13f7d35b.gif)

After:
![Screen Recording 2021-08-29 at 1 53 05 PM](https://user-images.githubusercontent.com/35318437/131265011-ea326c7d-0d51-4cf9-984a-13b0dfe875fb.gif)

Reply:
I dunno, I think it *should* close the overlay? Since the mouse down location was in the background. I'd even argue we could move it to `MouseDown` rather than `OnClick`?

Reply:
> I dunno, I think it should close the overlay? Since the mouse down location was in the background.

I have pushed a maybe "overly-complicated" test of what I expect. I feel like the background -> toolbar drag behavior should not close like background -> overlay.

> I'd even argue we could move it to MouseDown rather than OnClick?

Are you saying you want to change the feel of closing overlays to `OnMouseDown` and lose the drag behavior shown on the tests (i.e. user can prevent overlays from closing if they drag to overlay or toolbar?(this conversation))?

Reply:
I for one agree that moving to `MouseDown` sounds best. As in, holding down a mouse key outside of the overlay bounds would immediately close it, and there would be no chance to "salvage" the overlay being open via a drag anymore. We have other places that behave like this (i.e. popovers), and I'd rather have simple/idiomatic code over solving this one particular edge case the way it is done right now.

Reply:
Popovers have been recently added, and its closing behavior hasn't been really looked upon yet. While we have context menus and dropdown menus for like forever and they close `OnClick`/`OnMouseUp` when clicking a random place.

I thought of a better way of doing it (totally forgot about `ReceivePositionalInputAt`):
```diff
diff --git a/osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs b/osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs
index 0e635d26c2..68400d71aa 100644
--- a/osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs
+++ b/osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs
@@ -75,14 +75,14 @@ protected override void LoadComplete()
 
         protected override bool OnMouseDown(MouseDownEvent e)
         {
-            closeOnMouseUp = !base.ReceivePositionalInputAt(e.ScreenSpaceMousePosition) && (game?.Toolbar.IsHovered == false);
+            closeOnMouseUp = !base.ReceivePositionalInputAt(e.ScreenSpaceMousePosition) && game?.Toolbar.ReceivePositionalInputAt(e.ScreenSpaceMousePosition) == false;
 
             return base.OnMouseDown(e);
         }
 
         protected override void OnMouseUp(MouseUpEvent e)
         {
-            if (closeOnMouseUp && !base.ReceivePositionalInputAt(e.ScreenSpaceMousePosition) && (game?.Toolbar.IsHovered == false))
+            if (closeOnMouseUp && !base.ReceivePositionalInputAt(e.ScreenSpaceMousePosition) && game?.Toolbar.ReceivePositionalInputAt(e.ScreenSpaceMousePosition) == false)
                 Hide();
 
             base.OnMouseUp(e);
diff --git a/osu.Game/Overlays/Toolbar/Toolbar.cs b/osu.Game/Overlays/Toolbar/Toolbar.cs
index 918e3b7105..3d88171ba7 100644
--- a/osu.Game/Overlays/Toolbar/Toolbar.cs
+++ b/osu.Game/Overlays/Toolbar/Toolbar.cs
@@ -41,9 +41,6 @@ public class Toolbar : VisibilityContainer, IKeyBindingHandler<GlobalAction>
         // Toolbar and its components need keyboard input even when hidden.
         public override bool PropagateNonPositionalInputSubTree => true;
 
-        // IsHovered is used
-        public override bool HandlePositionalInput => true;
-
         public Toolbar()
         {
             RelativeSizeAxes = Axes.X;
@@ -143,13 +140,12 @@ public ToolbarBackground()
             protected override bool OnHover(HoverEvent e)
             {
                 gradientBackground.FadeIn(transition_time, Easing.OutQuint);
-                return base.OnHover(e);
+                return true;
             }
 
             protected override void OnHoverLost(HoverLostEvent e)
             {
                 gradientBackground.FadeOut(transition_time, Easing.OutQuint);
-                base.OnHoverLost(e);
             }
         }
```

Reply:
I don't honestly consider that much better than the current solution, at least not when it comes to code. What I'd ideally would like to eliminate from a code standpoint is the overlays being aware of / explicitly checking the state of the toolbar at all.

If the behaviour is to be stay then so be it, either variant of the code that you have is fine, but I really don't think this particular edge case being preserved is worth the detour.

Reply:
I found the “edge case” to be unexpected which is why I proposed the change. It’s weird logic and not how any OS works in sick cases.

Reply:
I can tell UI behaviors are really controversial and are not consistent per OS, website, or other application.

I can argue back to when I was implementing the opposite behavior:
> I'm not sure clicking anywhere on the toolbar should close overlays to be honest. The toolbar is considered at the same display level as the overlays

On Windows 10, open notifications then click the taskbar: notifications will close. There is also inconsistent behavior when opening the start menu and it won't close when clicking the taskbar (though on Windows 11, this has became consistent to the notifications behavior).

On macOS, open notifications then click the menu bar: notifications will close. 

At the end of the day, UI behaviors are really unexpected and overlooked upon. I can outright ignore this edge case and call it a quirk that I can get used to if no one else agrees.

Reply:
In osu!:

Things above other things should not close things below them.
The toolbar is above notifications. It should not close them.

Reply:
I'll let the "background -> toolbar" drag edge case slide, but I still disagree with changing the closing to `OnMouseDown` at this time. It should also be applied to context and dropdown menus if agreed upon but way out of scope of this PR.

Reply:
Should definitely be changed there too, yep. All operating systems to date do this.